url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/extensions/JudgmentCorrectionEquation.java

```diff
@@ -0,0 +1,212 @@
+	 * Note that the equation can be forced to be either assimilative or contrastive, but no both.*/
+	public boolean FORCE_CONTRASTIVE = GLOBAL_FORCE_CONTRASTIVE;
+	
+	/**
+	 * Initializes the judgment correction equation. The inputs of this equation must have two dimensions with the 
+	 * IDs: "PRIMER" and "TARGET" (of the enumerated type RequiredInputDimensions located in the JudgmentCorrectionModule). 
+	 * The values contained within these dimensions MUST have the same IDs as the values within the "RATING" (of the 
+	 * enumerated type RequiredActionDimensions located in the JudgmentCorrectionModule) dimension of the actions that are 
+	 * eligible for judgment correction. These IDs MUST be of a type that can be parsed into the primitive type 
+	 * double (e.g., Integer, Float, a numeric String, etc.).
+	 * <p>
+	 * The outputs for the equation are the actions that are to be involved in judgment correction.
+	 * @param InputSpace The input space for the equation.
+	 * @param Outputs The outputs for the equation.
+	 */
+	public JudgmentCorrectionEquation(Collection<Dimension> InputSpace,
+			AbstractOutputChunkCollection<? extends AbstractOutputChunk> Outputs) {
+		
+		super(InputSpace, Outputs);
+	}
+
+	/**
+     * Performs a forward pass from the input to output of the implicit module. This is the method used by
+     * the CLARION Library to obtain activations on the output layer given the input.
+	 * @throws SettingsConflictException If both FORCE_ASSIMILATIVE and FORCE_CONTRASTIVE are both set to true.
+	 */
+	public void forwardPass() throws SettingsConflictException{
+		
+		if(FORCE_ASSIMILATIVE && FORCE_CONTRASTIVE)
+			throw new SettingsConflictException ("The settings FORCE_ASSIMILATIVE and FORCE_CONTRASTIVE " +
+					"cannot both be set to true.");
+		
+		Dimension p = InputAsCollection.get(JudgmentCorrectionModule.RequiredInputDimensions.PRIMER);
+		Dimension t = InputAsCollection.get(JudgmentCorrectionModule.RequiredInputDimensions.TARGET);
+		
+		double Rp = 0;
+		for(Value v : p.values())
+		{
+			if(v.isFullyActivated())
+			{
+				Rp = Double.parseDouble(v.getID().toString());
+				break;
+			}
+		}
+		
+		double Rt = 0;
+		for(Value v : t.values())
+		{
+			if(v.isFullyActivated())
+			{
+				Rt = Double.parseDouble(v.getID().toString());
+				break;
+			}
+		}
+		
+		double correction = Math.abs(Rt - Rp);
+		if(FORCE_CONTRASTIVE || (!FORCE_ASSIMILATIVE && correction > PHI_1))
+		{
+			correction = Rt + (BETA_2 * Rp);
+		}
+		else if (FORCE_ASSIMILATIVE || (!FORCE_CONTRASTIVE && correction < PHI_2))
+		{
+			correction = Rt - (BETA_1 * Rp);
+		}
+		else
+		{
+			correction = Rt;
+		}
+		
+		double max = 0;
+		for(AbstractOutputChunk o : getOutput())
+		{
+			for(Value v : o.get(JudgmentCorrectionModule.RequiredActionDimensions.RATING).values())
+			{
+				double dv = Double.parseDouble(v.getID().toString());
+				
+				if(v.isFullyActivated() && dv > max)
+				{
+					max = dv;
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 157:168, 171:182.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.